src/Toolkit/PubSub/V1/PublisherApi.php

```diff
@@ -0,0 +1,673 @@
+     * a project resource.
+     */
+    public static function formatProjectName($project)
+    {
+        return self::getProjectNameTemplate()->render([
+            'project' => $project,
+        ]);
+    }
+
+    /**
+     * Formats a string containing the fully-qualified path to represent
+     * a topic resource.
+     */
+    public static function formatTopicName($project, $topic)
+    {
+        return self::getTopicNameTemplate()->render([
+            'project' => $project, 'topic' => $topic,
+        ]);
+    }
+
+    /**
+     * Parses the project from the given fully-qualified path which
+     * represents a project resource.
+     */
+    public static function parseProjectFromProjectName($projectName)
+    {
+        return self::getProjectNameTemplate()->match($projectName)['project'];
+    }
+
+    /**
+     * Parses the project from the given fully-qualified path which
+     * represents a topic resource.
+     */
+    public static function parseProjectFromTopicName($topicName)
+    {
+        return self::getTopicNameTemplate()->match($topicName)['project'];
+    }
+
+    /**
+     * Parses the topic from the given fully-qualified path which
+     * represents a topic resource.
+     */
+    public static function parseTopicFromTopicName($topicName)
+    {
+        return self::getTopicNameTemplate()->match($topicName)['topic'];
+    }
+
+    private static function getProjectNameTemplate()
+    {
+        if (self::$projectNameTemplate == null) {
+            self::$projectNameTemplate = new PathTemplate('projects/{project}');
+        }
+
+        return self::$projectNameTemplate;
+    }
+
+    private static function getTopicNameTemplate()
+    {
+        if (self::$topicNameTemplate == null) {
+            self::$topicNameTemplate = new PathTemplate('projects/{project}/topics/{topic}');
+        }
+
+        return self::$topicNameTemplate;
+    }
+
+    private static function getPageStreamingDescriptors()
+    {
+        $listTopicsPageStreamingDescriptor =
+                new PageStreamingDescriptor([
+                    'requestPageTokenField' => 'page_token',
+                    'responsePageTokenField' => 'next_page_token',
+                    'resourceField' => 'topics',
+                ]);
+        $listTopicSubscriptionsPageStreamingDescriptor =
+                new PageStreamingDescriptor([
+                    'requestPageTokenField' => 'page_token',
+                    'responsePageTokenField' => 'next_page_token',
+                    'resourceField' => 'subscriptions',
+                ]);
+
+        $pageStreamingDescriptors = [
+            'listTopics' => $listTopicsPageStreamingDescriptor,
+            'listTopicSubscriptions' => $listTopicSubscriptionsPageStreamingDescriptor,
+        ];
+
+        return $pageStreamingDescriptors;
+    }
+
+    // TODO(garrettjones): add channel (when supported in gRPC)
+    /**
+     * Constructor.
+     *
+     * @param array $options {
```

Comment:
Do you think it would be worthwhile to make sure the configuration options for toolkit/veneer code are the same?